### PR TITLE
MidiSynthExample: use N low latency APIs

### PR DESCRIPTION
**Pull request by philburk**
Optimize buffer size to get close to native latency.

This modification shows how to use the AudioAttributes.FLAG_LOW_LATENCY to reduce
latency. It also shows how to use AudioTrack.setBufferSizeInFrames() and getUnderrunCount() to tune the buffer for optimal latency.

### CODE REVIEW

File: MidiSynthExample/src/com/mobileer/midisynthexample/MidiSynthDeviceService.java [hunk 1]
```diff
@@ -16,20 +16,39 @@
...
+    }
+
+    // Query the system for the best sample rate and buffer size for low latency.
+    public void queryOptimalAudioSettings() {
+        AudioManager audioManager = (AudioManager) getSystemService(Context.AUDIO_SERVICE);
+     //   String text = audioManager.getProperty(AudioManager.PROPERTY_OUTPUT_SAMPLE_RATE);
+     //   mSampleRate = Integer.parseInt(text);
+     //TODO   mSynthEngine.setSampleRate(mSampleRate);
+        String text = audioManager.getProperty(AudioManager.PROPERTY_OUTPUT_FRAMES_PER_BUFFER);
+        mFramesPerBlock = 64; //TODO HACK - Integer.parseInt(text);
```
Comment (dturner) [inline MidiSynthExample/src/com/mobileer/midisynthexample/MidiSynthDeviceService.java]: This should either be fixed or removed.

File: MidiTools/src/com/mobileer/miditools/synth/SynthEngine.java [hunk 2]
```diff
@@ -36,8 +36,7 @@
 
     private static final String TAG = "SynthEngine";
 
-    public static final int FRAME_RATE = 48000;
-    private static final int FRAMES_PER_BUFFER = 240;
+    private static final int FRAMES_PER_BUFFER = 64;
```
Comment (dturner) [inline MidiTools/src/com/mobileer/miditools/synth/SynthEngine.java]: Why is this 64?

File: MidiTools/src/com/mobileer/miditools/synth/SynthEngine.java [hunk 3]
```diff
@@ -54,15 +53,23 @@
     private MidiFramer mFramer;
     private MidiReceiver mReceiver = new MyReceiver();
     private SimpleAudioOutput mAudioOutput;
+    private int mSampleRate = 48000;
```
Comment (dturner) [inline MidiTools/src/com/mobileer/miditools/synth/SynthEngine.java]: Shouldn't this be obtained using the AudioManager API?

File: MidiSynthExample/src/com/mobileer/midisynthexample/MidiSynthDeviceService.java (hunk 1 above)
Comment (philburk) [inline MidiSynthExample/src/com/mobileer/midisynthexample/MidiSynthDeviceService.java]: Thanks. I missed that hack. Will remove.

File: MidiTools/src/com/mobileer/miditools/synth/SynthEngine.java (hunk 3 above)
Comment (philburk) [inline MidiTools/src/com/mobileer/miditools/synth/SynthEngine.java]: Glenn says the rate may change before the track is opened. The best way to get the sample rate is to use an AudioTrack.Builder and not specify a rate. Then query the AudioTrack for the rate. I will do that.

File: MidiTools/src/com/mobileer/miditools/synth/SynthEngine.java (hunk 2 above)
Comment (philburk) [inline MidiTools/src/com/mobileer/miditools/synth/SynthEngine.java]: Now using framesPerBlock from AudioManager.

**Pull request merged**